msal/managed_identity.py

```diff
@@ -0,0 +1,598 @@
+        if payload.get("access_token") and payload.get("expires_on"):
+            return {  # Normalizing the payload into OAuth2 format
+                "access_token": payload["access_token"],
+                "expires_in": int(  # Despite the example in docs shows an integer,
+                    payload["expires_on"]  # Azure SDK team's test obtained a string.
+                    ) - int(time.time()),
+                "resource": payload.get("resource"),
+                "token_type": payload["token_type"],
+                }
+        error = payload.get("error", {})  # https://learn.microsoft.com/en-us/azure/service-fabric/how-to-managed-identity-service-fabric-app-code#error-handling
+        error_mapping = {  # Map Service Fabric errors into OAuth2 errors  https://www.rfc-editor.org/rfc/rfc6749#section-5.2
+            "SecretHeaderNotFound": "unauthorized_client",
+            "ManagedIdentityNotFound": "invalid_client",
+            "ArgumentNullOrEmpty": "invalid_scope",
+            }
+        return {
+            "error": error_mapping.get(payload["error"]["code"], "invalid_request"),
+            "error_description": resp.text,
+            }
+    except json.decoder.JSONDecodeError:
+        logger.debug("IMDS emits unexpected payload: %s", resp.text)
+        raise
+
+
+_supported_arc_platforms_and_their_prefixes = {
+    "linux": "/var/opt/azcmagent/tokens",
+    "win32": os.path.expandvars(r"%ProgramData%\AzureConnectedMachineAgent\Tokens"),
+}
+
+class ArcPlatformNotSupportedError(ManagedIdentityError):
+    pass
+
+def _obtain_token_on_arc(http_client, endpoint, resource):
+    # https://learn.microsoft.com/en-us/azure/azure-arc/servers/managed-identity-authentication
+    logger.debug("Obtaining token via managed identity on Azure Arc")
+    resp = http_client.get(
+        endpoint,
+        params={"api-version": "2020-06-01", "resource": resource},
+        headers={"Metadata": "true"},
+        )
+    www_auth = "www-authenticate"  # Header in lower case
+    challenge = {
+        # Normalized to lowercase, because header names are case-insensitive
+        # https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
+        k.lower(): v for k, v in resp.headers.items() if k.lower() == www_auth
+        }.get(www_auth, "").split("=")  # Output will be ["Basic realm", "content"]
+    if not (  # https://datatracker.ietf.org/doc/html/rfc7617#section-2
+            len(challenge) == 2 and challenge[0].lower() == "basic realm"):
+        raise ManagedIdentityError(
+            "Unrecognizable WWW-Authenticate header: {}".format(resp.headers))
+    if sys.platform not in _supported_arc_platforms_and_their_prefixes:
+        raise ArcPlatformNotSupportedError(
+            f"Platform {sys.platform} was undefined and unsupported")
+    filename = os.path.join(
+        _supported_arc_platforms_and_their_prefixes[sys.platform],
+        os.path.splitext(os.path.basename(challenge[1]))[0] + ".key")
```

Comment:
According to the example at https://learn.microsoft.com/en-us/azure/azure-arc/servers/managed-identity-authentication#acquiring-an-access-token-using-rest-api, shouldn't `challenge[1]` already be the secret file path?

```ps
$secretFile = ($wwwAuthHeader -split "Basic realm=")[1]
...
Write-Host "Secret file path: " $secretFile`n
```
![](https://learn.microsoft.com/en-us/azure/azure-arc/servers/media/managed-identity-authentication/powershell-token-output-example.png)

Why do we want to extract the filename then adding back the containing folder and file extension `.key`?

Reply:
Added a comment with an internal doc.